File: mybox/package/npm.py
import shlex
from typing import AsyncIterable

from pydantic import Field

from ..tracker import Tracker
from ..utils import allow_singular_none
from .base import Package
from .manual_version import ManualVersion
from .system import SystemPackage


class NpmPackage(ManualVersion):
    package: str = Field(..., alias="npm")

    binaries: list[str] = Field(default_factory=list, alias="binary")
    binaries_val = allow_singular_none("binaries")

    def derive_name(self) -> str:
        return self.package

    async def get_remote_version(self) -> str:
        try:
            result = await self.driver.run_output(
                "npm", "view", self.package, "version"
            )
        except Exception as exc:
            raise Exception(
                f"Cannot find latest version of package '{self.package}'."
            ) from exc
        version = result.strip()
        return version

    async def install(self, *, tracker: Tracker) -> None:
        args = ["npm", "exec", "--yes", "--package", self.package, "--"]

        await self.driver.run(*args, "true")

        npx_paths = await self.driver.run_output(*args, "echo $PATH")
        npx_path = next((path for path in npx_paths.split(":") if "_npx" in path), None)
        if not npx_path:
            raise Exception(
                f"Could not find npx path in {npx_paths}."
            )  # pragma: no cover

        for name in self.binaries:
            target = await self.driver.local() / "bin" / name
            await self.driver.write_file(
                target,
                f'#!/bin/sh\nPATH={shlex.quote(npx_path)}:$PATH\nexec "{name}" "$@"',
            )
            await self.driver.make_executable(target)
            tracker.track(target, root=self.root)

        await super().install(tracker=tracker)

    async def prerequisites(self) -> AsyncIterable[Package]:
        async for package in super().prerequisites():
            yield package  # pragma: no cover

        os = await self.driver.os()

        for system in os.switch_(
            linux=lambda linux: [
                "nodejs",
                {
                    "debian": "npm",
                    "ubuntu": "npm",
                    "fedora": "nodejs-npm",
                }[linux.distribution],
            ],
            macos=lambda: ["node"],
        ):
            yield SystemPackage(
                system=system,
                db=self.db,
                driver=self.driver_,
            )

File: src/Mybox/Package/NPM.hs
module Mybox.Package.NPM (NPMPackage (..), mkNPMPackage) where

import Data.Text qualified as Text

import Mybox.Aeson
import Mybox.Driver
import Mybox.Effects
import Mybox.Package.Class
import Mybox.Package.ManualVersion
import Mybox.Package.Post
import Mybox.Package.Queue
import Mybox.Package.System
import Mybox.Prelude
import Mybox.Tracker

data NPMPackage = NPMPackage
  { package :: Text
  , binaries :: [Text]
  , post :: [Text]
  }
  deriving (Eq, Show)

mkNPMPackage :: Text -> NPMPackage
mkNPMPackage package = NPMPackage{package, binaries = [], post = []}

instance HasField "name" NPMPackage Text where
  getField p = p.package

instance FromJSON NPMPackage where
  parseJSON = withObjectTotal "NPMPackage" $ do
    package <- takeField "npm"
    binaries <- takeCollapsedList "binary"
    post <- takePost
    pure NPMPackage{..}

instance ToJSON NPMPackage where
  toJSON p = object $ ["npm" .= p.package, "binary" .= p.binaries] <> postToJSON p

prerequisites :: App es => Eff es ()
prerequisites = do
  packages <- flip fmap drvOS $ \case
    Linux (Debian _) -> ["nodejs", "npm"]
    Linux Fedora -> ["nodejs", "nodejs-npm"]
    Linux (Generic d) -> terror $ "Cannot install npm on generic Linux: " <> d
    MacOS -> ["node"]
  for_ packages $ \package ->
    queueInstall $ mkSystemPackage package

viewVersion :: App es => NPMPackage -> Eff es Text
viewVersion p = do
  prerequisites
  drvRunOutput $ "npm" :| ["view", p.package, "version"]

npmInstall :: App es => NPMPackage -> Eff es ()
npmInstall p = do
  prerequisites

  let npmExec args = ("npm" :| ["exec", "--yes", "--package", p.package, "--"]) <> args

  -- Implicitly install the package
  drvRun $ npmExec ("true" :| [])

  -- Find the path where the binaries are installed
  npxPaths <- drvRunOutput $ npmExec $ shellRaw "echo $PATH"
  let npxBin' = find (elem "_npx" . (.segments)) $ map (mkPath @Abs) $ Text.splitOn ":" npxPaths

  case npxBin' of
    Nothing -> terror $ "Could not find npx path in " <> npxPaths
    Just npxBin -> do
      -- npxBin would be .../_npx/(hash)/node_modules/.bin, take the part up to
      -- and including the hash
      let npxPath = pParent npxBin >>= pParent
      for_ npxPath $ trkAdd p
      local <- drvLocal
      let binDir = local </> "bin"
      drvMkdir binDir

      forM_ p.binaries $ \name -> do
        let target = binDir </> name
        let script =
              Text.unlines
                [ "#!/bin/sh"
                , "PATH=" <> shellQuote npxBin.text <> ":$PATH"
                , "exec \"" <> name <> "\" \"$@\""
                ]
        drvWriteFile target script
        drvMakeExecutable target
        trkAdd p target

instance Package NPMPackage where
  localVersion = manualVersion
  remoteVersion = viewVersion
  install = installWithPost $ manualVersionInstall npmInstall
